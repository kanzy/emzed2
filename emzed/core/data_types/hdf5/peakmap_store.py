# encoding: utf-8
from __future__ import print_function, division

import itertools

from tables import Atom, UInt32Col, Float32Col, UInt8Col, UInt64Col, StringCol
import numpy as np

from emzed_optimizations import sample_peaks as optim_sample_peaks

from .. import PeakMap, Spectrum

from .store_base import Store, filters
from .lru import LruDict, lru_cache

from .install_profile import profile


class PeakMapStore(Store):

    ID_FLAG = 0
    HANDLES = PeakMap

    MSLEVEL_FIELD_SIZE = 16

    def __init__(self, file_, node, **kw):
        self.file_ = file_
        self.node = node
        self.setup()
        self.write_cache = LruDict(100)
        self.read_cache = LruDict(100)

    def setup(self):
        self.setup_blobs()
        self.setup_spec_table()
        self.setup_peakmap_table()

    def setup_blobs(self):
        if not hasattr(self.node, "mz_blob"):
            self.file_.create_earray(self.node, "mz_blob",
                                     Atom.from_dtype(np.dtype("float64")), (0,),
                                     filters=filters,
                                     )

        if not hasattr(self.node, "ii_blob"):
            self.file_.create_earray(self.node, "ii_blob",
                                     Atom.from_dtype(np.dtype("float32")), (0,),
                                     filters=filters,
                                     )

    def setup_spec_table(self):
        if not hasattr(self.node, "spec_table"):
            description = {}
            description["pm_index"] = UInt32Col(pos=0)
            description["rt"] = Float32Col(pos=1)
            description["ms_level"] = UInt8Col(pos=2)
            description["start"] = UInt64Col(pos=3)
            description["size"] = UInt32Col(pos=4)
            spec_table = self.file_.create_table(self.node, 'spec_table', description,
                                                 filters=filters)

            # every colums which appears in a where method call should/must be indexed !
            # this is not only for performance but for correct lookup as well (I had strange bugs
            # else)
            spec_table.cols.pm_index.create_index()
            spec_table.cols.rt.create_index()

    def setup_peakmap_table(self):
        if not hasattr(self.node, "pm_table"):
            description = {}
            description["unique_id"] = StringCol(itemsize=64, pos=0)
            description["index"] = UInt32Col(pos=1)
            description["ms_levels"] = StringCol(itemsize=self.MSLEVEL_FIELD_SIZE, pos=2)
            description["rtmin"] = Float32Col(pos=3)
            description["rtmax"] = Float32Col(pos=4)
            description["mzmin"] = Float32Col(pos=5)
            description["mzmax"] = Float32Col(pos=6)
            pm_table = self.file_.create_table(self.node, 'pm_table', description, filters=filters)
            # every colums which appears in a where method call should/must be indexed !
            # this is not only for performance but for correct lookup as well (I had strange bugs
            # else)
            pm_table.cols.unique_id.create_index()
            pm_table.cols.index.create_index()

    def add_spectrum(self, pm_index, spec):

        start = self.node.mz_blob.nrows
        self.node.mz_blob.append(spec.peaks[:, 0])
        self.node.ii_blob.append(spec.peaks[:, 1])

        size = spec.peaks.shape[0]

        row = self.node.spec_table.row
        row["pm_index"] = pm_index
        row["rt"] = spec.rt
        row["ms_level"] = spec.msLevel
        row["start"] = start
        row["size"] = size

        row.append()

    def _write(self, col_index, pm):

        # at the moment we ignore col_index, I guess it would not speed up so much

        unique_id = pm.uniqueId()
        # hash key
        yield unique_id

        result = list(self.node.pm_table.where("""unique_id == %r""" % unique_id))
        if result:
            yield result[0]["index"]
            return

        ms_levels = sorted(set(pm.getMsLevels()))
        as_str = ",".join(map(str, ms_levels))
        if len(as_str) > self.MSLEVEL_FIELD_SIZE:
            raise ValueError("ms level description %r is to long" % as_str)

        rtmin, rtmax = pm.rtRange()
        mzmin, mzmax = pm.mzRange()

        index = self.node.pm_table.nrows
        row = self.node.pm_table.row
        row["unique_id"] = unique_id
        row["index"] = index
        row["ms_levels"] = as_str
        row["rtmin"] = rtmin
        row["rtmax"] = rtmax
        row["mzmin"] = mzmin
        row["mzmax"] = mzmax
        row.append()

        for spec in pm.spectra:
            self.add_spectrum(index, spec)

        yield index

    def _read(self, col_index, index):
        result = list(self.node.pm_table.where("""index == %r""" % index))
        if len(result) == 0:
            raise ValueError("index %d not in table" % index)

        ms_levels = result[0]["ms_levels"]
        rtmin = result[0]["rtmin"]
        rtmax = result[0]["rtmax"]
        mzmin = result[0]["mzmin"]
        mzmax = result[0]["mzmax"]
        unique_id = result[0]["unique_id"]
        ms_levels = map(int, ms_levels.split(","))

        result = PeakMapProxy(node=self.node,
                              index=index,
                              ms_levels=ms_levels,
                              rtmin=rtmin,
                              rtmax=rtmax,
                              mzmin=mzmin,
                              mzmax=mzmax,
                              unique_id=unique_id)
        return result

    def flush(self):
        self.node.pm_table.flush()
        self.node.spec_table.flush()


class PeakMapProxy(object):

    def __init__(self, **kw):
        for name, value in kw.items():
            setattr(self, name, value)

        self.meta = {"unique_id": kw.get("unique_id")}

    def uniqueId(self):
        return self.unique_id

    def getMsLevels(self):
        return self.ms_levels

    def rtRange(self):
        return self.rtmin, self.rtmax

    def mzRange(self):
        return self.mzmin, self.mzmax

    def _iter_peaks(self, rtmin, rtmax, mzmin, mzmax, ms_level=1):
        if rtmin is not None and rtmax is not None:
            query = """(pm_index == %d) & (%f <= rt) & (rt <= %f)""" % (self.index, rtmin, rtmax)
        elif rtmin is not None:
            query = """(pm_index == %d) & (%f <= rt)""" % (self.index, rtmin)
        elif rtmax is not None:
            query = """(pm_index == %d) & (rt <= %f)""" % (self.index, rtmax)
        else:
            query = """(pm_index == %d)""" % (self.index,)
        mz_blob = self.node.mz_blob
        ii_blob = self.node.ii_blob
        for row in self.node.spec_table.where(query):
            if ms_level != row["ms_level"]:
                continue
            rt = row["rt"]
            i_start = row["start"]
            i_end = i_start + row["size"]
            mzs = mz_blob[i_start:i_end]
            iis = ii_blob[i_start:i_end]

            flags = True
            if mzmin is not None:
                flags = flags * (mzmin <= mzs)
            if mzmax is not None:
                flags = flags * (mzs <= mzmax)

            if flags is not True:
                idx = np.where(flags)[0]
                iis = iis[idx]
                mzs = mzs[idx]

            yield rt, mzs, iis

    @lru_cache(maxsize=1000)
    def chromatogram(self, mzmin=None, mzmax=None, rtmin=None, rtmax=None, ms_level=1):
        rts = []
        intensities = []
        for rt, mzs, iis in self._iter_peaks(rtmin, rtmax, mzmin, mzmax, ms_level):
            rts.append(rt)
            intensities.append(np.sum(iis))
        rts = np.array(rts)
        intensities = np.array(intensities)
        perm = np.argsort(rts)
        rts = rts[perm]
        intensities = intensities[perm]
        return rts, intensities

    @lru_cache(maxsize=1000)
    def sample_peaks(self, rtmin, rtmax, mzmin, mzmax, npeaks, ms_level):
        spectra = []
        for rt, mzs, iis in self._iter_peaks(rtmin, rtmax, mzmin, mzmax, ms_level):
            peaks = np.vstack((mzs, iis)).T
            spectrum = Spectrum(peaks, rt, ms_level, "0", [])
            spectra.append(spectrum)
        if rtmin <= rtmax and mzmin < mzmax:
            peaks = optim_sample_peaks(PeakMap(spectra),
                                       rtmin, rtmax, mzmin, mzmax, npeaks, ms_level)
        else:
            peaks = np.zeros((0, 2))
        return peaks

    def get_rts(self, msLevel=1):
        t = self.node.spec_table
        pm_indices = t.cols.pm_index[:]
        rts = t.cols.rt[:]
        ms_levels = t.cols.ms_level[:]
        return [r for (r, i, l) in itertools.izip(rts, pm_indices, ms_levels) if i == self.index and
                l == msLevel]
