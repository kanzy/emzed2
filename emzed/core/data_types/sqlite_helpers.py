# encoding: utf-8
from __future__ import print_function, division, absolute_import

import pickle
import itertools
import math
import os
import operator
import sqlite3
import time

from contextlib import contextmanager
from collections import namedtuple

from .table import Table
from .ms_types import PeakMap, PeakMapProxy


TableInformation = namedtuple("TableInformation", ["col_names", "col_types", "col_formats", "meta",
                                                   "object_columns"])


def _set_sort_permutation(conn, filter_expression, fields, ascending_indicators):

    c = conn.cursor()
    # c.execute('''DROP TABLE IF EXISTS mem.PERMUTATION_TABLE''')
    c.execute('''CREATE TABLE IF NOT EXISTS mem.PERMUTATION_TABLE (
                                        permutation INTEGER
                                        )''')

    q = c.execute("SELECT %s, ROWID FROM rows WHERE %s" % (", ".join(fields), filter_expression))
    rows = q.fetchall()
    for i, asc in list(enumerate(ascending_indicators))[::-1]:
        rows.sort(key=operator.itemgetter(i), reverse=not asc)
    perm = list((row[-1],) for row in rows)

    c.execute('''DELETE FROM mem.PERMUTATION_TABLE;''')

    c.executemany('''INSERT INTO mem.PERMUTATION_TABLE VALUES(?)''', perm)
    conn.commit()
    c.execute('''DROP INDEX IF EXISTS mem.permutation_index;''')
    c.execute('''CREATE INDEX mem.permutation_index on PERMUTATION_TABLE (permutation);''')
    conn.commit()

    select_stmt = '''SELECT rows.* FROM rows
                                   JOIN (SELECT ROWID, permutation FROM mem.PERMUTATION_TABLE) as P
                                   ON P.permutation = rows.ROWID
                                   ORDER BY P.ROWID'''
    c.close()
    return select_stmt


@contextmanager
def timeit(msg):
    started = time.time()
    yield
    needed = time.time() - started
    print("%s needed %e seconds" % (msg, needed))


def _setup_fresh_db(db_name="peaks.db"):
    if os.path.exists(db_name):
        os.remove(db_name)

    conn = sqlite3.connect(db_name)
    return conn


def _table_names(conn):
    names = conn.execute("""SELECT name FROM sqlite_master WHERE type='table';""").fetchall()
    return [n[0] for n in names]


class Sqlite3ObjectProxy(object):

    def __init__(self, id_, name):
        self.id_ = id_
        self.name = name

    def __str__(self):
        return self.name


class Sqlite3PeakMapProxy(Sqlite3ObjectProxy):

    pass


def _pickle_peakmap(obj, conn):
    names = _table_names(conn)
    next_id = len([n for n in names if n.startswith("ms1_spectra__")])
    table_name = "ms1_spectra__%d" % next_id

    cursor = conn.cursor()

    stmt = """CREATE TABLE %s (
                    rt INTEGER, mz FLOAT, iie INTEGER, iifrac INTEGER
              )""" % table_name

    cursor.execute(stmt)

    stmt = "INSERT INTO %s VALUES(?, ?, ?, ?)""" % table_name

    for spectrum in obj:
        if spectrum.msLevel == 1:
            rt = int(round(spectrum.rt * 10))
            for (mz, ii) in spectrum.peaks:
                iie = int(math.log10(ii))
                iifrac = int((ii / 10 ** iie) * 1000)
                cursor.execute(stmt, (rt, mz, iie, iifrac))

    cursor.execute("CREATE INDEX rt_index_%d_ms1 on %s(rt);" % (next_id, table_name))
    cursor.execute("CREATE INDEX mz_index_%d_ms1 on %s(mz);" % (next_id, table_name))

    conn.commit()

    cursor = conn.cursor()
    table_name = "ms2_spectra__%d" % next_id
    stmt = """CREATE TABLE %s (
                    rt INTEGER, mz FLOAT, iie INTEGER, iifrac INTEGER,
                    precursor FLOAT
              )""" % table_name

    cursor.execute(stmt)
    cursor.execute("CREATE INDEX rt_index_%d_ms2 on %s(rt);" % (next_id, table_name))
    cursor.execute("CREATE INDEX mz_index_%d_ms2 on %s(mz);" % (next_id, table_name))
    cursor.execute("CREATE INDEX precursor_index_%d on %s(precursor);" % (next_id, table_name))
    cursor = conn.cursor()

    stmt = "INSERT INTO %s VALUES(?, ?, ?, ?, ?)""" % table_name

    for spectrum in obj:
        if spectrum.msLevel == 2:
            precursor = spectrum.precursors[0][0]
            rt = int(round(spectrum.rt * 10))
            for (mz, ii) in spectrum.peaks:
                iie = int(math.log10(ii))
                iifrac = int((ii / 10 ** iie) * 1000)
                cursor.execute(stmt, (rt, mz, iie, iifrac, precursor))

    conn.commit()

    return Sqlite3PeakMapProxy(next_id, str(obj))


def _pickle_object(obj, conn):

    if type(obj) in (PeakMap, PeakMapProxy):
        return _pickle_peakmap(obj, conn)

    cursor = conn.cursor()
    name = str(type(obj))
    cursor.execute("""INSERT INTO blobs VALUES(NULL, ?, ?); """,
                   (name, buffer(pickle.dumps(obj, protocol=pickle.HIGHEST_PROTOCOL)),))

    proxy = Sqlite3ObjectProxy(cursor.lastrowid, repr(obj))
    return proxy


def load_object(id_, conn):
    stmt = "SELECT name, blob FROM blobs WHERE id = ?"
    name, blob = conn.cursor().execute(stmt, (id_,)).fetchone()
    return name, blob


def unpickle_object(id_, conn):
    name, blob = load_object(id_, conn)
    return name, pickle.loads(blob)


def to_db(t, name):
    conn = _setup_fresh_db(name)
    try:
        _to_db(t, conn)
    except Exception:
        conn.close()
        raise
    finally:
        conn.close()


def get_name(cell):
    return str(type(cell))


def _to_db(t, conn):

    type_map = {int: "INTEGER", float: "REAL", str: "TEXT", unicode: "TEXT",
                bool: "INTEGER"}

    declarations = []
    object_columns = []
    for name, type_ in zip(t._colNames, t._colTypes):

        sql_type = type_map.get(type_)
        if sql_type is None:
            declarations.append(("__id_%s" % name, "TEXT"))
            object_columns.append(name)
        else:
            declarations.append(("%s" % name, sql_type))

    col_definitions = ",\n".join("'%s' %s" % name_type_pair for name_type_pair in declarations)

    stmt = """CREATE TABLE rows (
                 %s
              );
    """ % col_definitions

    conn.cursor().execute(stmt)
    conn.commit()

    stmt = """CREATE TABLE blobs (
                 id INTEGER PRIMARY KEY,
                 name TEXT,
                 blob BLOB
              );"""

    print(stmt)
    conn.cursor().execute(stmt)
    conn.commit()

    info = TableInformation(t._colNames, t._colTypes, t._colFormats, t.meta, object_columns)
    id_ = _pickle_object(info, conn).id_
    assert id_ == 1

    object_cache = {}
    n_cols = len(t._colNames)

    cursor = conn.cursor()

    for row in t.rows:

        stmt = """INSERT INTO rows VALUES(%s)""" % ", ".join(["?"] * n_cols)

        values = [None]
        values = []
        for cell, type_ in zip(row, t._colTypes):
            if type_ in type_map:
                if cell is None:
                    value = None
                else:
                    value = type_(cell)
            else:
                if cell is None:
                    value = None
                else:
                    value = object_cache.get(id(cell))
                    if value is None:
                        obj = _pickle_object(cell, conn)
                        value = "%d!%s" % (obj.id_, obj.name)
                        object_cache[id(cell)] = value
            values.append(value)

        cursor.execute(stmt, values)

    cursor.close()
    conn.commit()

    for name, __ in declarations:
        stmt = """CREATE INDEX '%s_index' ON rows ('%s'); """ % (name, name)
        conn.cursor().execute(stmt)
        conn.commit()


def print_meta(db_name):
    conn = sqlite3.connect(db_name)
    name, info = _unpickle_object(1, conn)
    for name, t, f in zip(info.col_names, info.col_types, info.col_formats):
        print(name, t, f)
    print(info)


def from_db(db_name, n_rows=None):

    conn = sqlite3.connect(db_name)
    name, info = unpickle_object(1, conn)

    # fetch all rows
    stmt = """SELECT * from rows;"""
    if n_rows is None:
        rows = map(list, conn.cursor().execute(stmt).fetchall())
    else:
        rows = map(list, conn.cursor().execute(stmt).fetchmany(n_rows))

    # resolve pickled objects
    cache = {}

    def lookup(id_):
        if id_ is None:
            return None
        if id_ not in cache:
            cache[id_] = unpickle_object(id_, conn)
        return cache[id_]

    meta_rows = conn.cursor().execute("""PRAGMA table_info(rows);""").fetchall()

    db_col_names = [str(meta_row[1]) for meta_row in meta_rows]

    for i, name in enumerate(db_col_names):
        if name.startswith("__id_"):
            info.col_names[i] = name[5:]
            for row in rows:
                row[i], obj = lookup(row[i])

    return Table(info.col_names, info.col_types, info.col_formats, rows, meta=info.meta)


def _sort_permutation(db_name, col_names):

    stmt = """SELECT ROWID FROM rows ORDER BY %s""" % (", ".join(col_names))
    conn = sqlite3.connect(db_name)
    with timeit("compute permutation to order by %s with SQL   " % (", ".join(col_names))):
        permutation = [r[0] for r in conn.cursor().execute(stmt).fetchall()]
        print(len(permutation))

    stmt = """SELECT %s FROM rows""" % (", ".join(col_names))
    conn = sqlite3.connect(db_name)
    with timeit("compute permutation to order by %s with Python" % (", ".join(col_names))):
        decorated = [r + (i,) for i, r in enumerate(conn.cursor().execute(stmt).fetchall())]
        decorated.sort()
        permutation = [d[-1] for d in decorated]
        print(len(permutation))


def filter(db_name, filters, col_names):

    stmt = """SELECT ROWID FROM rows ORDER BY %s""" % (", ".join(col_names))
    conn = sqlite3.connect(db_name)

    with timeit("compute permutation to order by %s with SQL   " % (", ".join(col_names))):
        permutation = [r[0] for r in conn.cursor().execute(stmt).fetchall()]
        print(len(permutation))

    stmt = """SELECT %s FROM rows""" % (", ".join(col_names))
    conn = sqlite3.connect(db_name)
    with timeit("compute permutation to order by %s with Python" % (", ".join(col_names))):
        decorated = [r + (i,) for i, r in enumerate(conn.cursor().execute(stmt).fetchall())]
        decorated.sort()
        permutation = [d[-1] for d in decorated]
        print(len(permutation))


def fetch_column(db_name, col_name, blocksize=None):
    conn = sqlite3.connect(db_name)
    conn.commit()
    stmt = """SELECT %s FROM rows order by ROWID""" % col_name
    conn = sqlite3.connect(db_name)
    if blocksize is None:
        for (r,) in conn.cursor().execute(stmt).fetchall():
            yield r
    else:
        for block in conn.cursor().execute(stmt).fetchmany(blocksize):
            for r in block:
                yield r


def add_column(db_name, col_name, type_, values):
    conn = sqlite3.connect(db_name)
    c = conn.cursor()
    started = time.time()

    c.execute('''ATTACH DATABASE ':memory:' AS memory;''')

    c.execute('''DROP TABLE IF EXISTS memory.TEMP''')
    c.execute('''CREATE TABLE memory.TEMP (v %s)''' % type_)
    conn.commit()
    print (time.time() - started)

    c.execute("PRAGMA synchronous=OFF")
    c.executemany('''INSERT INTO memory.TEMP VALUES(?)''', ((v,) for v in values))

    conn.commit()
    print (time.time() - started)
    c.execute('''ALTER TABLE rows ADD COLUMN %s %s;''' % (col_name, type_))
    conn.commit()
    print (time.time() - started)
    c.execute("PRAGMA synchronous=NORMAL")
    c.execute('''UPDATE rows SET %s = (SELECT v FROM memory.TEMP)''' % col_name)
    conn.commit()
    print (time.time() - started)
    c.execute('''CREATE INDEX %s_index on rows(%s)''' % (col_name, col_name))
    conn.commit()
    print (time.time() - started)
    return (time.time() - started)


if __name__ == "__main__":

    if True:
        t = Table.loadTable("peak_table.table")

        def generate(i):
            if i % 3 == 0:
                return None
            elif i % 3 == 1:
                return ""
            return "x"
        t.addColumn("empty", t.peak_id.apply(generate), type_=str)
        t.addColumn("sample_id", 0, type_=int)

        o0 = [range(i, j) for i in range(0, 10) for j in range(10, 20)]
        o1 = [set(range(i, j)) for i in range(0, 10) for j in range(10, 20)]
        o2 = [dict(zip(range(i, j), range(i, j))) for i in range(0, 10) for j in range(10, 20)]
        o3 = [o0, o1, o2]
        o4 = [[o3, [o3]], o3]
        o5 = [o0, o1, o2, o3, o4]

        objects = [o0, o1, o2, o3, o4, o5]

        def by_6(x):
            return x % 6

        t.addColumn("objects", t.peak_id.apply(by_6).apply(objects.__getitem__), type_=object)

        tables = []
        for i in range(500):
            if i % 10 == 0:
                print(i)
            t.replaceColumn("sample_id", i, type_=int)
            tables.append(t[:])
            break
        t = Table.stackTables(tables)
        to_db(t, "peaks.db")
        t.info()

    elif 0:

        values = fetch_column("peaks.db", "peak_id")
        new_values = (v + 2.0 for v in values)
        add_column("peaks.db", "peak_id_plus_twoxxxf", "FLOAT", new_values)
    elif 0:
        key = None
        pi = sort_permutation("peaks.db", "mz <= 101.1", ["peak_id", "sample_id"], [False, True])
        pi = sort_permutation("peaks.db", "mz <= 151.1", ["peak_id", "sample_id"], [False, True])
    else:
        # benchmark deserialization
        tsub = from_db("peaks.db", 20)
        conn = sqlite3.connect("peaks.db")
        s = time.time()
        v = fetch_column("peaks.db", "__id_objects", blocksize=500)
        tunp = 0.0
        s = time.time()
        n = 10
        for vi in itertools.islice(v, n):
            st = time.time()
            name, o = unpickle_object(vi, conn)
            # eval(str(o))
            tunp += time.time() - st
        print("overall fetch", n, "objects:", time.time() - s)
        print("unpickle time is", tunp)

    if 0:
        tn = from_db("peaks.db")
        tn.info()

        assert t._colNames == tn._colNames
        assert t._colTypes == tn._colTypes
        assert t._colFormats == tn._colFormats
        assert t.meta == tn.meta
        assert t.uniqueId() == tn.uniqueId()
