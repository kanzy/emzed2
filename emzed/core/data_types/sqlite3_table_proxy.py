# encoding: utf-8
from __future__ import print_function, division, absolute_import

import operator
import sqlite3


from .sqlite_helpers import unpickle_object, chromatogram


class Sqlite3TableProxy(object):

    def __init__(self, path):
        self.conn = sqlite3.connect(path)
        self.conn.execute('''ATTACH DATABASE ':memory:' AS mem''')
        self.read_table_info()

        self.sort_order = ""
        self.select_for_mixed_sort_order = None
        self.reset_filter()
        self.cursor = None

    def read_table_info(self):
        __, info = unpickle_object(1, self.conn)

        for (name, value) in info._asdict().items():
            setattr(self, name, value)

    def reset_filter(self):
        self.filter_expression = "0=0"

    def create_query(self):
        stmt = "SELECT * FROM rows WHERE {filter_} {order}".format(filter_=self.filter_expression,
                                                                   order=self.sort_order)
        if self.cursor is not None:
            self.cursor.close()
        print(stmt)
        self.cursor = self.conn.cursor().execute(stmt)
        return stmt

    def fetchone(self):
        assert self.cursor is not None, "call create_query() method first"
        return self.cursor.fetchone()

    def set_sort_order(self, items):
        assert all(len(item) == 2 for item in items)

        if len(items) == 0:
            self.sort_order = ""
            return

        term = ", ".join("%s %s" % (name, "ASC" if ascending else "DESC") for (name, ascending)
                         in items)
        self.sort_order = "ORDER BY %s" % term

    def set_filter(self, filter_expression):
        self.filter_expression = filter_expression

    def check_filter_expression(self, filter_expression):
        stmt = "SELECT * FROM rows WHERE (%s) LIMIT 1" % filter_expression
        cursor = self.conn.cursor()
        try:
            cursor.execute(stmt)
        except Exception as e:
            return str(e)
        return None

    def get_chromatogram(self, rtmin, rtmax, mzmin, mzmax, peakmap_entry):
        peakmap_id = int(peakmap_entry.split("!")[0])
        return chromatogram(self.conn, peakmap_id, rtmin, rtmax, mzmin, mzmax)
