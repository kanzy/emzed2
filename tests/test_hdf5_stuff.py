# encoding: utf-8, division
from __future__ import print_function, division

import os
import tempfile

from datetime import datetime

import numpy as np

from emzed.core.data_types import TimeSeries
from emzed.core.data_types.hdf5_table_writer import to_hdf5, append_to_hdf5
from emzed.core.data_types.hdf5_table_proxy import Hdf5TableProxy, ObjectProxy, PeakMapProxy
from emzed.core.data_types.ms_types import PeakMap, Spectrum

from emzed.core.data_types.hdf5.accessors import (Hdf5TableWriter, Hdf5TableAppender,
                                                  Hdf5TableReader)


from emzed.utils import toTable

from pytest import fixture, raises, yield_fixture


@fixture()
def table():

    peaks_0 = np.arange(20.0).reshape(-1, 2)
    peaks_1 = np.arange(30.0).reshape(-1, 2)

    spec_0 = Spectrum(peaks_0[:], 30.0, 1, "+")
    spec_1 = Spectrum(peaks_1[:], 30.0, 2, "+", [(10, 1000)])
    spec_2 = Spectrum(peaks_1[:], 40.0, 1, "+")

    pm = PeakMap([spec_0, spec_1, spec_2])

    t0 = toTable("int", (1, 2, 3, 3, None), type_=int)
    t0.addColumn("float", (1.0, 2.0, 4.0, 3.0, None), type_=float)
    t0.addColumn("bool", (True, False, None, True, False), type_=bool)
    t0.addColumn("str", ("1", "2" * 100, None, "a", "b"), type_=str)
    t0.addColumn("object", ({1}, dict(a=2), None, (1,), [1, 2]), type_=object)
    t0.addColumn("peakmap", pm, type_=object)

    n = 10
    ts_0 = TimeSeries(map(datetime.fromordinal, range(1, n + 1)), range(n), "label1")

    n = 10
    ts_1 = TimeSeries(map(datetime.fromordinal, range(1000,  n + 1000)),
                      range(1000, 1000 + n), "label2", [bool(i % 4) for i in range(n)])

    t0.addColumn("time_series", [ts_0, ts_0, ts_1, ts_1, None], type_=TimeSeries)

    t0.meta = {"a": "b"}
    return t0


@yield_fixture()
def tproxy(table):

    folder = tempfile.mkdtemp()
    path = os.path.join(folder, "t.hdf5")
    to_hdf5(table, path)

    tproxy = Hdf5TableProxy(path)
    yield tproxy
    tproxy.close()


def test_writer_appender_reader(table, tmpdir, regtest):

    path = tmpdir.join("test.hdf5").strpath
    writer = Hdf5TableWriter(path)
    writer.write_table(table)
    writer.close()

    appender = Hdf5TableAppender(path)
    appender.append_table(table)
    appender.append_table(table)
    appender.close()

    reader = Hdf5TableReader(path)
    assert len(reader) == 3 * len(table)
    rows = list(reader)

    # resove proxies:
    rows = [[ci.load() if isinstance(ci, ObjectProxy) else ci for ci in row] for row in rows]

    # check if appending worked:
    assert rows[:5] == rows[5:10]
    assert rows[5:10] == rows[10:]
    reader.close()


def test_hdf5_table_appender(table, tmpdir):
    path = tmpdir.join("test.hdf5").strpath
    to_hdf5(table, path)
    append_to_hdf5([table, table, table], path)


def test_round_trip(tproxy, table, regtest):
    # test roundtrip:
    t1 = tproxy.toTable()
    assert t1.uniqueId() == table.uniqueId()
    assert t1.meta == table.meta

    for pm0, pm1 in zip(t1.peakmap, table.peakmap):
        assert isinstance(pm0, PeakMapProxy)
        assert isinstance(pm1, PeakMap)
        rt0, ii0 = pm0.chromatogram(0, 1000, 0, 1000)   # fetch lazy
        rt1, ii1 = pm1.chromatogram(0, 1000, 0, 1000)   # data is in memory
        assert np.all(rt0 == rt1)
        assert np.all(ii0 == ii1)

    print(t1, file=regtest)
    return


def test_get_index(tproxy):

    # test getIndex method
    assert tproxy.getIndex("int") == 0
    assert tproxy.getIndex("object") == 4

    with raises(Exception):
        tproxy.getIndex("hih")


def test_row_attribute(tproxy, regtest):
    # test row attribute
    for row in tproxy.rows:
        print(row, file=regtest)


def test_row_write_cell(tproxy, regtest):
    # test row attribute
    tproxy.setCellValue(0, 0, 4711)

    t = tproxy.toTable()
    t.dropColumns("time_series", "object", "peakmap")
    print(t, file=regtest)

    # make sure that row 0 is cached, and test again:
    tproxy[0]
    tproxy.setCellValue(0, 0, 4712)
    t = tproxy.toTable()
    t.dropColumns("time_series", "object", "peakmap")
    print(t, file=regtest)

    tproxy.setCellValue(4, 0, 4713)
    t = tproxy.toTable()
    t.dropColumns("time_series", "object", "peakmap")
    print(t, file=regtest)

    tproxy.setCellValue(4, 0, None)
    t = tproxy.toTable()
    t.dropColumns("time_series", "object", "peakmap")
    print(t, file=regtest)

    tproxy.setCellValue(4, 0, 4713)
    t = tproxy.toTable()
    t.dropColumns("time_series", "object", "peakmap")
    print(t, file=regtest)

    tproxy.setCellValue(4, 0, None)
    t = tproxy.toTable()
    t.dropColumns("time_series", "object", "peakmap")
    print(t, file=regtest)

    tproxy.setCellValue(0, 3, "4713")
    t = tproxy.toTable()
    t.dropColumns("time_series", "object", "peakmap")
    print(t, file=regtest)

    tproxy.setCellValue(0, 3, None)
    t = tproxy.toTable()
    t.dropColumns("time_series", "object", "peakmap")
    print(t, file=regtest)

def test_sort_by(tproxy, regtest):
    # test sortBy
    perm = tproxy.sortPermutation("int", True)
    print(perm, file=regtest)
    print(tproxy.toTable()[perm], file=regtest)

    perm = tproxy.sortPermutation("int", False)
    print(perm, file=regtest)
    print(tproxy.toTable()[perm], file=regtest)

    perm = tproxy.sortPermutation(("int", "float"), (True, True))
    print(perm, file=regtest)
    print(tproxy.toTable()[perm], file=regtest)

    perm = tproxy.sortPermutation(("int", "float"), (True, False))
    print(perm, file=regtest)
    print(tproxy.toTable()[perm], file=regtest)


def test_ghost_table_forwarding(tproxy):
    assert tproxy.getValue(tproxy.rows[1], "int") == 2

    assert tproxy.supportedPostfixes("") == []
