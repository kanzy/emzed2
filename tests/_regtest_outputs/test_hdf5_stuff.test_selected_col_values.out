[1]
[1, 2]
[1, 2, 3]
[1, 2, 3, 3]
[1, 2, 3, 3, None]
[2, 3, 3, None]
[3, 3, None]
[3, None]
[None]
